libqrexec/toml.c

```diff
@@ -0,0 +1,275 @@
+            LOG(INFO, "%s:%zu:missing newline at EOF", config_full_path, lineno);
+        }
+        // Multi-line strings not yet implented, so just chop off trailing whitespace
+        while (linelen > 1 && qubes_isspace(current_line[linelen - 1])) {
+            linelen--;
+            current_line[linelen] = '\0';
+        }
+        switch (current_line[0]) {
+        case '\0':
+        case '#':
+            // Skip comments and blank lines
+            continue;
+        case 'a' ... 'z':
+        case 'A' ... 'Z':
+            break;
+        case '[':
+            LOG(ERROR, "%s:%zu: TOML section headers not supported", config_full_path, lineno);
+            goto bad;
+        case ' ':
+        case '\t':
+            LOG(ERROR, "%s:%zu: Unexpected whitespace at start of line", config_full_path, lineno);
+            goto bad;
+        default:
+            if (current_line[0] > ' ' && current_line[0] <= '~') {
+                LOG(ERROR, "%s:%zu: Invalid character '%c' at start of key", config_full_path, lineno, current_line[0]);
+            } else {
+                LOG(ERROR, "%s:%zu: Invalid byte 0x%x at start of key", config_full_path, lineno, current_line[0]);
+            }
+            goto bad;
+        }
+        unsigned char *key_cursor = (unsigned char *)current_line;
+        do {
+            key_cursor++;
+        } while (qubes_is_key_byte(key_cursor[0]));
+        int const key_len = key_cursor - (unsigned char *)current_line;
+        while (qubes_isspace(key_cursor[0]))
+            key_cursor++;
+        if (key_cursor[0] != '=') {
+            LOG(ERROR, "%s:%zu: Missing '=' after key", config_full_path, lineno);
+            goto bad;
+        }
+        do {
+            key_cursor++;
+        } while (qubes_isspace(key_cursor[0]));
+        current_line[key_len] = '\0';
+        union toml_data value;
+        enum toml_type ty = parse_toml_value(config_full_path, lineno, key_cursor, &value);
+
+        if (strcmp(current_line, "wait-for-session") == 0) {
+            if (seen_wait_for_session) {
+                LOG(ERROR, "%s:%zu: Key '%s' appears more than once", config_full_path, lineno, current_line);
+                goto bad;
+            }
+            seen_wait_for_session = true;
+            if (ty == TOML_TYPE_BOOL) {
+                *wait_for_session = (int)value.boolean;
+                continue;
+            } else if (ty == TOML_TYPE_INTEGER) {
+                if (value.integer < 2) {
+                    *wait_for_session = (int)value.integer;
+                    continue;
+                }
+                LOG(ERROR, "Integer value %llu used when a boolean was expected", value.integer);
+            } else if (ty == TOML_TYPE_STRING) {
+                LOG(ERROR, "String value '%s' not valid for 'wait-for-session'", value.string);
+                free(value.string);
+                value.string = NULL;
+            }
+        } else if (strcmp(current_line, "user") == 0) {
```

Comment:
Maybe better call it `force-user`, since it will override `qvm-run --user` argument?